Changes to data structures

mplsmitch opened the pull request:
#### **What problem does your proposal solve? Please begin with the relevant issue number. If there is no existing issue, please also describe alternative solutions you have considered.**
We have two fields,`vehicle_type_area_capacity` and `vehicle_type_dock_capacity` that use data from other fields as object keys. This is inconsistent with other similar fields in the specification such as `vehicle_types_available` and has caused some challenges in generating models and validation. 

#### **What is the proposal?**
This proposal changes the objects to arrays as was done with `vehicle_types _available`. I've read through the rest of the specification and haven't found this issue elsewhere. If anyone is aware other examples please chime in so we can roll them into this PR.

#### **Is this a breaking change?**
- [x] Yes 
- [ ] No
- [ ] Unsure

#### **Which files are affected by this change?**

mplsmitch committed:
Objects to arrays

Changes objects vehicle_type_dock_capacity and vehicle_type_area_capacity to arrays.

mplsmitch committed:
formatting fixes

formatting fixes to JSON examples

mplsmitch committed:
Merge branch 'master' into data-types-02022022

testower commented:
Entur supports this change

testower commented:
@mplsmitch How can we move this forward? Call to vote?

mplsmitch commented:
Hello @testower - the next step is to call a vote on this along with #454, #457 and #460 and anything else we'd like to get into the v3.0 release. In the past we've made the most progress if we group the votes an a number of issues, as opposed to holding individual votes. The PRs need to be open for comment for 7 days before a vote can be called.

I'd like to do something about #306, basically it would be like your file: https://api.entur.io/mobility/v2/gbfs/
I should be able to write it up today and get a PR open.  

  Great work on the Internationalization!

josee-sabourin commented:
I hereby call a vote on this proposal. Voting will be open for 10 full calendar days until 11:59PM UTC on Friday, January 20th.

Please vote for or against the proposal, and include the organization for which you are voting in your comment.

Please note if you can commit to implementing the proposal.

cmonagle commented:
+1 from Transit

testower commented:
+1 from Entur

bdferris-v2 commented on gbfs.md:
```diff
@@ -625,11 +629,17 @@ Field Name | REQUIRED | Type | Defines
-          "abc123": 7,
-          "def456": 9
-        }
+        "is_charging_station": "true",
```
did you mean to stringify `true` here?

josee-sabourin replied:
Probably a mistake - will fix when/if we merge.

bdferris-v2 replied:
Dumb question, but why not fix it now?  Isn't it better to be precise on what we are voting for?

bdferris-v2 commented on gbfs.md:
```diff
@@ -595,11 +595,15 @@ Field Name | REQUIRED | Type | Defines
-\-&nbsp;`vehicle_type_area_capacity` <br/>*(added in v2.1)* | OPTIONAL | Object | An object used to describe the parking capacity of virtual stations (defined using the `is_virtual_station` field), where each key is a `vehicle_type_id` as described in [vehicle_types.json](#vehicle_typesjson) and the value is a number representing the total number of vehicles of this type that can park within the  ...
-\-&nbsp;`vehicle_type_dock_capacity` <br/>*(added in v2.1)* | OPTIONAL | Object | An object used to describe the docking capacity of a station where each key is a `vehicle_type_id` as described in [vehicle_types.json](#vehicle_typesjson) and the value is a number representing the total docking points installed at this station, both available and unavailable for the specified vehicle type.
+\-&nbsp;`vehicle_type_area_capacity` <br/>*(as of v3.0)* | OPTIONAL | Array| This field's value is an array of objects containing the keys `vehicle_type_id` and `count` defined below.  These objects are used to model the parking capacity of virtual stations (defined using the `is_virtual_station` field) for each vehicle type defined in `vehicle_types.json`.
+&emsp;&emsp;\-&nbsp;`vehicle_type_id`| Conditionally REQUIRED | ID | A `vehicle_type_id`, as defined in `vehicle_types.json`, that may park at the virtual station.
```
Dumb question, but why is this "Conditionally Required"?  Under what conditions would it be left empty?  Same for `count` field below.

josee-sabourin replied:
They are required _**if**_ `vehicle_type_area_capacity` is used, if not, they are not required. We can clarify this when/if we merge.

bdferris-v2 replied:
What is the convention in the spec for a field that is required if its parent element is specified, where the parent element may be optional?  If I understand correctly, in this PR the required child fields are being marked as "Conditionally Required".  However, I see in PR #460 that newly added child fields are being marked simply as "Required", even if there parent element is optional.  I'll admit the latter convention seems more natural to me, but understood if that's not the convention used in the spec.

mplsmitch replied:
Sorry for the delayed response, I've been traveling and just got back. I'll update the PR with corrections.  @bdferris-v2 you have a good point, we need to be consistent on how parent-child relationships are described throughout. I think you are correct that they should be labeled simply as required in this case because we could have parent-child relationships with other dependencies making them conditionally required.  Making these consistent throughout the spec wouldn't change the behavior of any fields so it won't require a vote. We should just go ahead and do it now so it can get rolled into the next release.

Some day I want to write a style guide that would contain all these conventions.

mplsmitch committed:
Change child field requirements, remove string (#475)

Change child fields of optional parent from Conditionally Required to Required.  Change incorrect string to boolean.
Needed for PR #457

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MobilityData/gbfs?pullRequest=407) <br/>All committers have signed the CLA.

josee-sabourin commented:
Voting on this PR closes in 2 calendar days. Please vote for or against the proposal, and include the organization for which you are voting in your comment. Please note if you can commit to implementing the proposal.

bdferris-v2 commented on gbfs.md:
```diff
@@ -595,11 +595,15 @@ Field Name | REQUIRED | Type | Defines
+&emsp;&emsp;\-&nbsp;`count`| Yes | Non-negative integer | REQUIRED if `vehicle_type_area_capacity` is defined. A number representing the total number of vehicles of the corresponding `vehicle_type_id` type that can park within the virtual station.<br /><br />If the virtual station is defined by `station_area`, this is the number that can park within the station area. If `lat`/`lon` is defined, th ...
+\-&nbsp;`vehicle_type_dock_capacity` <br/>*(as of v3.0)* | OPTIONAL | Array | This field's value is an array of objects containing the keys `vehicle_type_id` and `count` defined below. These objects are used to model the total docking capacity of a station, both available and unavailable, for each type of vehicle defined in `vehicle_types.json`.
+&emsp;&emsp;\-&nbsp;`vehicle_type_id`| Yes | ID | REQUIRED if `vehicle_type_id` is defined. A `vehicle_type_id`, as defined in `vehicle_types.json`, that may dock at the station.
+&emsp;&emsp;\-&nbsp;`count`| Yes | Non-negative integer | REQUIRED if `vehicle_type_id` is defined. The total number of docks at the station, both available and unavailable, that may accept the corresponding vehicle type as defined by its `vehicle_type_id`.
```
I don't think your follow-up change was quite right.  Shouldn't this be:

> REQUIRED if `vehicle_type_dock_capacity` is defined.

Both here and for the line above?

Though in generally, I think you could get away with deleting all "REQUIRED if `parent_element` is defined." statements and just be explicit about the behavior of a required field with an optional parent in the spec preamble.

mplsmitch replied:
oops! thanks for catching that- fixed it

mplsmitch committed:
Fix typo in vehicle_type_dock_capacity (#478)

* Change child field requirements, remove string

Change child fields of optional parent from Conditionally Required to Required.  Change incorrect string to boolean.

* fix typo in vehicle_type_dock_capacity

rootsixtysix commented:
+1 for harmonization of data structure from nextbike

kheraankit commented:
+1 form Lime

benwedge commented:
+1 from Joyride

cait32 commented:
+1 from BCycle

josee-sabourin commented:
This vote has now closed, and it passes!

Votes in favor:
Transit (consumer)
Entur (consumer)
Nextbike (producer)
Lime (producer)
Joyride (producer)
BCycle (producer)
 
There were no votes against.
Thank you to everyone who took the time to review and to vote on this! We incorporate it into 3.0-RC, which should be ready to go in the coming week.
